docs/developer/decision-records/2023-02-28-processing-callbacks/README.md

```diff
@@ -0,0 +1,147 @@
+- The optional `states` property is a collection of processing states corresponding to a state machine state. If absent, the default value is all states. The '*' character may be
+  used as a wildcard to explicitly denote all states.
+- If the optional `transactional` property is true, callback dispatches will be conducted transactionally (on supported configurations) during the state machine transition. The
+  default is false.
+- The optional `auth-key` property defines the transport specific header to use for sending an authentication value. If the `auth-code-id` value is set and this property is not
+  set, the value `Authorization` will be used.
+- The optional id to use when resolving a required callback authorization key from the Vault.
+
+Static callback endpoints are referenced by their id when a client makes a request. They may be reused by multiple requests.
+
+### Dynamic Endpoints
+
+Dynamic endpoints are registered as part of a client request and are therefore scoped to the latter. Dynamic endpoints have three properties:
+
+- `uri`
+- `context`
+- `states`
+- `transactional`
+
+Dynamic endpoints do not have explicit API keys. Security can be provided at the network layer or through a URI with an randomly-generated single-use path part.
+
+## New Services
+
+### The CallbackRegistry
+
+A `CallbackRegistry` to manage static endpoints will be added to the `control-plane-spi` module:
+
+```
+package org.eclipse.edc.connector.spi.callback
+
+public interface CallbackRegistry {
+
+  void register(CallbackAddress address);
+  
+  List<CallbackAddress> resolve(String context, String state);
+
+}
+```
+
+This registry will be implemented in `control-plane-aggregate-services` as a simple, in-memory service. Dynamic endpoint will be persisted with the transfer process.
+
+## Impact on Existing Services
+
+### Management API
+
+`ContractNegotiationDto` and `TransferRequestDto` will be enhanced to include a collection of `CallbackAddress` types that model dynamic endpoint configurations:
+
+```
+public class CallbackAddress {
+
+    private String uri;
+    
+    private String context;
+    
+    private Set<String> states;
+
+    private boolean transactional 
+
+    ....
+}
+```
+
+Provided dynamic callback addresses will be persisted alongside the request.
+
+### State Machines
+
+During a state transition, a `ContractNegotiationListener` or `TransferProcessListener` will resolve `CallbackAddress` entries matching the state to be transitioned to by querying
+the `CallbackRegistry` and current `TransferProcess`. If a resolved callback is marked as transactional, an invocation error will mark the current transaction as rollback-only;
+otherwise, invocation errors will be logged and the transaction will proceed. Invocations will use the standard EDC retry mechanism.
+
+> Note that the EventRouter cannot be used for dispatching as the `ContractNegotiation` and `TransferProcess` must be available during dispatch
```

Comment:
We could move these into events. However, we would need to do two things:

1. Cleanup the event hierarchy and payload (based on previous discussions)
2. Pass most of the transfer process data to the event

I like you suggestion below of aligning the API with events instead of states so maybe we should stage work that would arise out of this DR after the event cleanup?

Reply:
We agreed to do these items in a separate PR that will be done prior to the callback work